creator-node/src/snapbackSM.js

```diff
@@ -47,63 +58,308 @@ class SnapbackSM {
+        // Enqueue sync for secondary1 if required
+        if (secondary1SyncRequired) {
+          await this.issueSecondarySync(userWallet, secondary1, this.endpoint)
+        }
+        // Enqueue sync for secondary2 if required
+        if (secondary2SyncRequired) {
+          await this.issueSecondarySync(userWallet, secondary2, this.endpoint)
+        }
+      }
+    ))
+    this.log('------------------END Process SnapbackSM Operation------------------')
   }
 
-  /*
-    Initialize the configs necessary to run
-  */
+  // Track an ongoing sync operation to a secondary
+  async monitorSecondarySync (syncWallet, primaryClockValue, secondaryUrl) {
+    // Monitor the sync status
+    let syncMonitoringRequestParameters = {
+      method: 'get',
+      baseURL: secondaryUrl,
+      url: `/sync_status/${syncWallet}`,
+      responseType: 'json'
+    }
+    let startTime = Date.now()
+    let syncAttemptCompleted = false
+    while (!syncAttemptCompleted) {
+      try {
+        let syncMonitoringResp = await axios(syncMonitoringRequestParameters)
+        let respData = syncMonitoringResp.data.data
+        this.log(`processSync ${syncWallet} secondary response: ${JSON.stringify(respData)}`)
+        // A success response does not necessarily guarantee completion, validate response data to confirm
+        if (respData.clockValue === primaryClockValue) {
+          syncAttemptCompleted = true
+          this.log(`processSync ${syncWallet} clockValue from secondary:${respData.clockValue}, primary:${primaryClockValue}`)
+        }
+      } catch (e) {
+        this.log(`processSync ${syncWallet} error querying sync_status: ${e}`)
+      }
+      if (Date.now() - startTime > MaxSyncMonitoringDurationInMs) {
+        this.log(`ERROR: processSync ${syncWallet} timeout for ${syncWallet}`)
+        syncAttemptCompleted = true
+      }
+      // 1s delay between retries
+      await utils.timeout(1000)
+    }
+  }
+
+  // Main sync queue job
+  async processSyncOperation (job) {
+    const syncRequestParameters = job.data.syncRequestParameters
+    let isValidSyncJobData = (
+      ('baseURL' in syncRequestParameters) &&
+      ('url' in syncRequestParameters) &&
+      ('method' in syncRequestParameters) &&
+      ('data' in syncRequestParameters)
+    )
+    if (!isValidSyncJobData) {
+      logger.error(`Invalid sync data found`)
+      logger.error(job.data)
+      return
+    }
+    const syncWallet = syncRequestParameters.data.wallet[0]
+    const primaryClockValue = await this.getUserPrimaryClockValue(syncWallet)
+    const secondaryUrl = syncRequestParameters.baseURL
+    this.log(`------------------Process SYNC | User ${syncWallet} | Target: ${secondaryUrl} ------------------`)
+
+    // Issue sync request to secondary
+    await axios(syncRequestParameters)
+
+    // Monitor the sync status
+    await this.monitorSecondarySync(syncWallet, primaryClockValue, secondaryUrl)
```

Comment:
so we basically keep retrying until we confirm successful sync or timeout. i'm concerned that if some users are in corrupted state and syncs keep failing we won't have visibility into that. i guess we could check loggly for al the error messages of `processSync ${syncWallet} error querying sync_status: ${e}` and perhaps build a quick dashboard around it

do you think thats sufficient?

Reply:
max timeout handles this nothing will live forever

Reply:
yes